toolchain/check/cpp_type_mapping.cpp

```diff
@@ -0,0 +1,179 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/cpp_type_mapping.h"
+
+#include <cstddef>
+#include <iostream>
+#include <optional>
+
+#include "clang/AST/Type.h"
+#include "clang/Basic/TargetInfo.h"
+#include "toolchain/base/int.h"
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/base/value_ids.h"
+#include "toolchain/check/context.h"
+#include "toolchain/check/convert.h"
+#include "toolchain/check/literal.h"
+#include "toolchain/sem_ir/ids.h"
+#include "toolchain/sem_ir/inst.h"
+#include "toolchain/sem_ir/type.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::Check {
+
+// Find the bit width of an integer literal.
+// The default bit width is 32. If the literal's bit width is greater than 32,
+// the bit width is increased to 64. Negative literals are assigned a bit width
+// of 64.
+static auto FindIntLiteralBitWidth(Context& context, SemIR::InstId arg_id,
+                                   bool is_signed) -> IntId {
+  auto bit_width_id = IntId::MakeRaw(32);
+  auto arg = context.insts().TryGetAs<SemIR::IntValue>(arg_id);
+  if (arg) {
+    auto arg_val = context.ints().Get(arg->int_id);
+
+    auto width = bit_width_id.has_value()
+                     ? context.ints().Get(bit_width_id).getZExtValue()
+                     : arg_val.getBitWidth();
+
+    unsigned arg_non_sign_bits = arg_val.getSignificantBits() - 1;
+    return (arg_non_sign_bits + is_signed <= width) ? bit_width_id
+                                                    : IntId::MakeRaw(64);
+  }
+  return IntId::MakeRaw(64);
+}
+
+// Maps a Carbon record type to a Cpp type.
+// Returns `std::nullopt` if the Carbon type is not a ClassType or if the
+// Cpp record type has not yet been imported.
+static auto TryMapRecordType(Context& context, SemIR::TypeId type_id)
+    -> std::optional<clang::QualType> {
+  auto class_type =
+      context.sem_ir().types().TryGetAs<SemIR::ClassType>(type_id);
+  if (!class_type) {
+    return std::nullopt;
+  }
+  const auto& class_info = context.sem_ir().classes().Get(class_type->class_id);
+  auto clang_decl_id =
+      context.name_scopes().Get(class_info.scope_id).clang_decl_context_id();
+  if (!clang_decl_id.has_value()) {
+    return std::nullopt;
+  }
+  clang::Decl* clang_decl =
+      context.sem_ir().clang_decls().Get(clang_decl_id).decl;
+  auto* record_type_decl = clang::dyn_cast<clang::CXXRecordDecl>(clang_decl);
+  return context.ast_context().getRecordType(record_type_decl);
```

Comment:
This will crash for enum types.
```suggestion
  auto* record_type_decl = clang::dyn_cast<clang::TagDecl>(clang_decl);
  return context.ast_context().getTagType(record_type_decl);
```
You should also handle the case where `dyn_cast` returns `nullptr`. (Maybe a CHECK? Or use `cast` instead of `dyn_cast`.)

Reply:
Done.